src/workflow/ArcanistPatchWorkflow.php

```diff
@@ -435,18 +435,18 @@ public function run() {
             $param);
           break;
         case self::SOURCE_DIFF:
+          $bundle = $this->loadDiffBundleFromConduit(
+             $this->getConduit(),
+             $param);
           if ($this->shouldMergeUsingStagingGitTag()) {
-            $bundle = $this->loadDiffBundleFromConduit(
-              $this->getConduit(),
-              $param);
             $this->mergeBranchFromStagingArea($param, $bundle);
             return 0;
           } elseif ($this->shouldUseStagingGitTags()) {
+            $repository_api = $this->getRepositoryAPI();
+            if (!$repository_api->hasLocalCommit($bundle->getBaseRevision())) {
               $this->pullBaseTagFromStagingArea($param);
             }
-          $bundle = $this->loadDiffBundleFromConduit(
-            $this->getConduit(),
-            $param);
```

Comment:
Doesn't it require getting $bundle after pull?

Reply:
loadDIffBundleFromConduit doesn't depend on local state of repository and pulling ref from repository will also not change any state in Phabricator itself. Maybe will waste a bit of cpu cycles if pullBaseTag fails but if it doesn't fail which is 99%+ cases - nothing much changed just duplication removal...
```
  final protected function loadDiffBundleFromConduit(
    ConduitClient $conduit,
    $diff_id) {

    return $this->loadBundleFromConduit(
      $conduit,
      array(
      'ids' => array($diff_id),
    ));
  }

  final protected function loadRevisionBundleFromConduit(
    ConduitClient $conduit,
    $revision_id) {

    return $this->loadBundleFromConduit(
      $conduit,
      array(
      'revisionIDs' => array($revision_id),
    ));
  }

  final private function loadBundleFromConduit(
    ConduitClient $conduit,
    $params) {

    $future = $conduit->callMethod('differential.querydiffs', $params);
    $diff = head($future->resolve());

    if ($diff == null) {
      throw new Exception(
        phutil_console_wrap(
          pht("The diff or revision you specified is either invalid or you ".
          "don't have permission to view it."))
      );
    }

    $changes = array();
    foreach ($diff['changes'] as $changedict) {
      $changes[] = ArcanistDiffChange::newFromDictionary($changedict);
    }
    $bundle = ArcanistBundle::newFromChanges($changes);
    $bundle->setConduit($conduit);
    // since the conduit method has changes, assume that these fields
    // could be unset
    $bundle->setBaseRevision(idx($diff, 'sourceControlBaseRevision'));
    $bundle->setRevisionID(idx($diff, 'revisionID'));
    $bundle->setAuthorName(idx($diff, 'authorName'));
    $bundle->setAuthorEmail(idx($diff, 'authorEmail'));
    return $bundle;
  }
```

Reply:
@mindaugasrukas - are you satisfied with answer?